Guard against kubernetes not being installed

mjpieters opened the pull request:
If the `kubernetes.client` import fails, then `airflow.kubernetes.pod_generator` also can't be imported, and there won't be attributes on `k8s` to use in `isinstance()` calls.

Instead of setting `k8s` to `None`, use an explicit flag so later code can disable kubernetes-specific branches explicitly.

closes: #11556

mjpieters committed:
Guard against kubernetes not being installed

If the `kubernetes.client` import fails, then `airflow.kubernetes.pod_generator` also can't be imported, and there won't be attributes on `k8s` to use in `isinstance()` calls.

Instead of setting `k8s` to `None`, use an explicit flag so later code can disable kubernetes-specific branches explicitly.

potiuk reviewed (commented):
LGTM. But It would be great to have unit test.

kaxil reviewed (approved):
LGTM

mjpieters committed:
Give isort some control but keep these two lines together.

ashb commented on airflow/serialization/serialized_objects.py:
```diff
@@ -46,6 +39,16 @@
+    # isort: on
+    has_kubernetes = True
+except ImportError:
+    has_kubernetes = False
```
```suggestion
    HAS_KUBERNETES = True
except ImportError:
    HAS_KUBERNETES = False
```

ashb commented on airflow/serialization/serialized_objects.py:
```diff
@@ -187,12 +190,12 @@ def _serialize(cls, var: Any) -> Any:  # Unfortunately there is no support for r
                     type_=DAT.DICT
                 )
-            elif isinstance(var, k8s.V1Pod):
+            elif has_kubernetes and isinstance(var, k8s.V1Pod):
```
```suggestion
            elif HAS_KUBERNETES and isinstance(var, k8s.V1Pod):
```

ashb commented on airflow/serialization/serialized_objects.py:
```diff
@@ -187,12 +190,12 @@ def _serialize(cls, var: Any) -> Any:  # Unfortunately there is no support for r
             elif isinstance(var, list):
                 return [cls._serialize(v) for v in var]
-            elif isinstance(var, k8s.V1Pod):
+            elif has_kubernetes and isinstance(var, k8s.V1Pod):
```
```suggestion
            elif HAS_KUBERNETES and isinstance(var, k8s.V1Pod):
```

ashb commented on airflow/serialization/serialized_objects.py:
```diff
@@ -256,7 +259,7 @@ def _deserialize(cls, encoded_var: Any) -> Any:  # pylint: disable=too-many-retu
         elif type_ == DAT.DATETIME:
             return pendulum.from_timestamp(var)
-        elif type_ == DAT.POD:
+        elif has_kubernetes and type_ == DAT.POD:
```
```suggestion
        elif HAS_KUBERNETES and type_ == DAT.POD:
```

ashb replied:
Though I also wonder if it should be an error to try and deserialise a pod without the kube modules installed?

So something like:

```suggestion
        elif type_ == DAT.POD:
            if not HAS_KUBERNETES:
                raise RuntimeError("Cannot deserialize POD objects without kubernetes libraries installed!")
```

mjpieters committed:
Use uppercase for detected feature

Also, when de-serializing a Kubernetes pod with no kubernetes library installed is an error.

mjpieters committed:
Add a unit test for optional kubernetes support

mjpieters committed:
Use names that don't look like built-ins

This makes pylint happier.

mjpieters commented on airflow/serialization/serialized_objects.py:
```diff
@@ -187,12 +190,12 @@ def _serialize(cls, var: Any) -> Any:  # Unfortunately there is no support for r
             elif isinstance(var, list):
                 return [cls._serialize(v) for v in var]
-            elif isinstance(var, k8s.V1Pod):
+            elif HAS_KUBERNETES and isinstance(var, k8s.V1Pod):
```
This block is redundant actually; no need to test for k8s.V1Pod _twice_.

kaxil replied:
PR: https://github.com/apache/airflow/pull/11602
